### ISSUE TITLE
Document truncated log occur in diags.log

### ISSUE THREAD

**Issue by taoyunxing** [apache/trafficserver#1906]
I always found some logs  in diags.log  on production host like this:

May 13 10:52:59.948] Server {0x7ffff3c2c700} WARNING: Document 6E175543E1C4B188B8DDD553EEF9862F truncated

I investigate this and finally filter a piece of url from thousands of requests of ATS 5.3.2/6.2.1, which can reappear this above log stably：
http://xunniu.vod.taocdn.readtv.cn/NewVideo/xunniu/2017011919/12129220.ts

I request this url from ATS in debug mode，
curl -vx 127.0.0.1:80 -o /dev/null 'http://xunniu.vod.taocdn.readtv.cn/NewVideo/xunniu/2017011919/12129220.ts'

I found the following：

[Switching to Thread 0x7ffff2313700 (LWP 72185)]

Breakpoint 2, CacheVC::openReadStartHead (this=0x7ffec801e9a0, event=3900, e=0x0) at CacheRead.cc:1037
1037	    if (doc->magic != DOC_MAGIC) {
(gdb) p *doc
$28 = {magic = 1595054867, len = 2904, total_len = 703308, first_key = {b = {6581159149356253959, 2363591363087607973}, u64 = {6581159149356253959, 2363591363087607973}, u32 = {3148940039, 1532295520,
      1728358565, 550316498}, u8 = "\a\003\261\273`\371T[\245\250\004g\322)\315 "}, key = {b = {12674803048354180674, 9498882061998288457}, u64 = {12674803048354180674, 9498882061998288457}, u32 = {
      4012928578, 2951082551, 2092756553, 2211630824}, u8 = "Bn0\357\067\362\345\257I\356\274|\350\316҃"}, hlen = 2832, doc_type = 3, v_major = 24, v_minor = 0, unused = 0, sync_serial = 34118,
  write_serial = 2149452, pinned = 0, checksum = 2695938256}
(gdb) c
Continuing.

Breakpoint 1, CacheVC::openReadStartEarliest (this=0x7ffec801e9a0) at CacheRead.cc:823
823	    if (doc->magic != DOC_MAGIC) {
(gdb) p *doc
$29 = {magic = 1595054867, len = 524288, total_len = 564741, first_key = {b = {6581159149356253959, 2363591363087607973}, u64 = {6581159149356253959, 2363591363087607973}, u32 = {3148940039, 1532295520,
      1728358565, 550316498}, u8 = "\a\003\261\273`\371T[\245\250\004g\322)\315 "}, key = {b = {9545684039016667831, 13240539856947867084}, u64 = {9545684039016667831, 13240539856947867084}, u32 = {
      3954465463, 2222527758, 935816652, 3082803417}, u8 = "\267Z\264\353\016\025y\204\314m\307\067\331ؿ\267"}, hlen = 0, doc_type = 3, v_major = 24, v_minor = 0, unused = 0, sync_serial = 34118,
  write_serial = 2149452, pinned = 0, checksum = 2695938256}
(gdb) c
Continuing.

Breakpoint 4, CacheVC::openReadMain (this=0x7ffec801e9a0) at CacheRead.cc:638
638	  int64_t ntodo = vio.ntodo();
(gdb) p *doc
$30 = {magic = 1595054867, len = 524288, total_len = 564741, first_key = {b = {6581159149356253959, 2363591363087607973}, u64 = {6581159149356253959, 2363591363087607973}, u32 = {3148940039, 1532295520,
      1728358565, 550316498}, u8 = "\a\003\261\273`\371T[\245\250\004g\322)\315 "}, key = {b = {9545684039016667831, 13240539856947867084}, u64 = {9545684039016667831, 13240539856947867084}, u32 = {
      3954465463, 2222527758, 935816652, 3082803417}, u8 = "\267Z\264\353\016\025y\204\314m\307\067\331ؿ\267"}, hlen = 0, doc_type = 3, v_major = 24, v_minor = 0, unused = 0, sync_serial = 34118,
  write_serial = 2149452, pinned = 0, checksum = 2695938256}
(gdb) c
Continuing.

Breakpoint 4, CacheVC::openReadMain (this=0x7ffec801e9a0) at CacheRead.cc:638
638	  int64_t ntodo = vio.ntodo();
(gdb) p *doc
$31 = {magic = 1595054867, len = 524288, total_len = 564741, first_key = {b = {6581159149356253959, 2363591363087607973}, u64 = {6581159149356253959, 2363591363087607973}, u32 = {3148940039, 1532295520,
      1728358565, 550316498}, u8 = "\a\003\261\273`\371T[\245\250\004g\322)\315 "}, key = {b = {9545684039016667831, 13240539856947867084}, u64 = {9545684039016667831, 13240539856947867084}, u32 = {
      3954465463, 2222527758, 935816652, 3082803417}, u8 = "\267Z\264\353\016\025y\204\314m\307\067\331ؿ\267"}, hlen = 0, doc_type = 3, v_major = 24, v_minor = 0, unused = 0, sync_serial = 34118,
  write_serial = 2149452, pinned = 0, checksum = 2695938256}
(gdb) c
Continuing.

Breakpoint 3, CacheVC::openReadReadDone (this=0x7ffec801e9a0, event=3900, e=0x0) at CacheRead.cc:559
559	      if (doc->magic != DOC_MAGIC) {
(gdb) p *doc
$32 = {magic = 1595054867, len = 179164, total_len = 703308, first_key = {b = {6581159149356253959, 2363591363087607973}, u64 = {6581159149356253959, 2363591363087607973}, u32 = {3148940039, 1532295520,
      1728358565, 550316498}, u8 = "\a\003\261\273`\371T[\245\250\004g\322)\315 "}, key = {b = {11108024346135017448, 13654429118677323557}, u64 = {11108024346135017448, 13654429118677323557}, u32 = {
      1002747880, 2586288458, 1837305637, 3179169520}, u8 = "\350\267\304;J\241'\232%\017\203m\360F~\275"}, hlen = 0, doc_type = 3, v_major = 24, v_minor = 0, unused = 0, sync_serial = 34118,
  write_serial = 2149452, pinned = 0, checksum = 2695938256}
(gdb) c
Continuing.

Breakpoint 4, CacheVC::openReadMain (this=0x7ffec801e9a0) at CacheRead.cc:638
638	  int64_t ntodo = vio.ntodo();
(gdb) p *doc
$33 = {magic = 1595054867, len = 179164, total_len = 703308, first_key = {b = {6581159149356253959, 2363591363087607973}, u64 = {6581159149356253959, 2363591363087607973}, u32 = {3148940039, 1532295520,
      1728358565, 550316498}, u8 = "\a\003\261\273`\371T[\245\250\004g\322)\315 "}, key = {b = {11108024346135017448, 13654429118677323557}, u64 = {11108024346135017448, 13654429118677323557}, u32 = {
      1002747880, 2586288458, 1837305637, 3179169520}, u8 = "\350\267\304;J\241'\232%\017\203m\360F~\275"}, hlen = 0, doc_type = 3, v_major = 24, v_minor = 0, unused = 0, sync_serial = 34118,
  write_serial = 2149452, pinned = 0, checksum = 2695938256}
(gdb) c
Continuing.

In my oppion, _total_len_ is object total length, fragment size is configured as 512KB, so each fragment is 524288bytes. This multi-fragment has 3 fragments, the first fragment just save header info (72 bytes Doc, plus 2832 hlen ) and no payload data, the earliest fragment save 524216 bytes payload and 72 bytes Doc(sizeofDoc=72), the last fragment just save 179092 bytes payload and 72 bytes Doc. _len_ is actual total length of each fragment. 

So，there are the following relation:
first fragment：         2904     = sizeofDoc(72) + hlen(2832)
earliest fragment:      524288 = sizeofDoc(72) + payload(524216)
last fragment:            179164  = sizeofDoc(71)  + payload(179092)
payload total length  703308 = 524216 + 179092

why the _total_len_ can vary from Doc of the fragments belonging to a multi-fragment? the total_len of Doc in earliest fragment should be 703308, not 564471 !
I deem this is a cache bug, I wish someone can help me!

**Comment by taoyunxing** [comment]
curl -vx 127.0.0.1:80 -o /dev/null 'http://xunniu.vod.taocdn.readtv.cn/NewVideo/xunniu/2017011919/12129220.ts'
* About to connect() to proxy 127.0.0.1 port 80 (#0)
*   Trying 127.0.0.1... connected
* Connected to 127.0.0.1 (127.0.0.1) port 80 (#0)
> GET http://xunniu.vod.taocdn.readtv.cn/NewVideo/xunniu/2017011919/12129220.ts HTTP/1.1
> User-Agent: curl/7.19.7 (x86_64-redhat-linux-gnu) libcurl/7.19.7 NSS/3.16.2.3 Basic ECC zlib/1.2.3 libidn/1.18 libssh2/1.4.2
> Host: xunniu.vod.taocdn.readtv.cn
> Accept: */*
> Proxy-Connection: Keep-Alive
>
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:06 --:--:--     0< HTTP/1.1 200 OK
< Server: nginx/1.9.15
< Date: Mon, 15 May 2017 10:11:39 GMT
< Content-Type: video/mp2t
< Content-Length: 703308
< Connection: keep-alive
< Last-Modified: Thu, 19 Jan 2017 07:12:15 GMT
< ETag: "a2f61410f4938507017c410c37b85f58"
< Expires: Sat, 20 May 2017 02:17:36 GMT
< Cache-Control: max-age=2592000
< Age: 2188511
< Via: http/1.1 Front-ZJLSCTC260 (TaoCacheBox V2.32 [cRs f ]), http/1.1 EDG-JSHACTC14 (TaoCacheBox V2.32 [cRs f ])
< Tao_Status: TCP_HIT
< Proxy-Connection: keep-alive
< Accept-Ranges: bytes
<
{ [data not shown]
 74  686k   74  511k    0     0  31326      0  0:00:22  0:00:16  0:00:06     0* transfer closed with 179575 bytes remaining to read
 74  686k   74  511k    0     0  31326      0  0:00:22  0:00:16  0:00:06     0* Closing connection #0

curl: (18) transfer closed with 179575 bytes remaining to read

As long as we trigger the above request to ATS, we can find a piece of log in diags.log like this:
[May 13 10:52:59.948] Server {0x7ffff3c2c700} WARNING: Document 6E175543E1C4B188B8DDD553EEF9862F truncated

**Comment by taoyunxing** [comment]
this following is the internal procedure of ATS on the debug mode：

[Switching to Thread 0x7ffff3020700 (LWP 35645)]

Breakpoint 1, CacheVC::openReadMain (this=0x7ffef001b360) at CacheRead.cc:638
638	  int64_t ntodo = vio.ntodo();
(gdb) c
Continuing.

Breakpoint 1, CacheVC::openReadMain (this=0x7ffef001b360) at CacheRead.cc:638
638	  int64_t ntodo = vio.ntodo();
(gdb) n
639	  int64_t bytes = doc->len - doc_pos;
(gdb) p ntodo
$54 = 179092
(gdb) n
640	  IOBufferBlock *b = NULL;
(gdb)
641	  if (seek_to) { // handle do_io_pread
(gdb)
715	  if (ntodo <= 0)
(gdb)
717	  if (vio.buffer.writer()->max_read_avail() > vio.buffer.writer()->water_mark && vio.ndone) // initiate read of first block
(gdb)
719	  if ((bytes <= 0) && vio.ntodo() >= 0)
(gdb)
720	    goto Lread;
(gdb)
740	  if (vio.ndone >= (int64_t)doc_len)
(gdb)
743	  last_collision = 0;
(gdb)
744	  writer_lock_retry = 0;
(gdb)
749	  cancel_trigger();
(gdb)
750	  CACHE_TRY_LOCK(lock, vol->mutex, mutex->thread_holding);
(gdb)
751	  if (!lock.is_locked()) {
(gdb)
755	  if (dir_probe(&key, vol, &dir, &last_collision)) {
(gdb)
756	    SET_HANDLER(&CacheVC::openReadReadDone);
(gdb)
757	    int ret = do_read_call(&key);
(gdb)
758	    if (ret == EVENT_RETURN)
(gdb)
790	  return handleEvent(AIO_EVENT_DONE, 0);
(gdb) c
Continuing.

Breakpoint 2, CacheVC::openReadReadDone (this=0x7ffef001b360, event=3900, e=0x0) at CacheRead.cc:559
559	      if (doc->magic != DOC_MAGIC) {
(gdb) n
573	      if (doc->key == key)
(gdb) p *doc
$55 = {magic = 1595054867, len = 3016, total_len = 4194304, first_key = {b = {7936001979168651200, 1986278447822286449}, u64 = {7936001979168651200, 1986278447822286449}, u32 = {3643078592, 1847744448,
      2071342705, 462466489}, u8 = "\300\367$\331\300W\"nq.v{\271\255\220\033"}, key = {b = {5241550030087605862, 11374957274309961022}, u64 = {5241550030087605862, 11374957274309961022}, u32 = {
      1070550630, 1220393467, 3061687614, 2648438623}, u8 = "fN\317?\373\271\275H>\245}\266_\367۝"}, hlen = 2944, doc_type = 3, v_major = 24, v_minor = 0, unused = 0, sync_serial = 34143,
  write_serial = 851889, pinned = 0, checksum = 2695938256}
(gdb) n
576	      else if (dir_ininterim(&dir)) {
(gdb)
577	        dir_delete(&key, vol, &dir);
(gdb)
578	        last_collision = NULL;
(gdb)
583	    if (last_collision && dir_get_offset(&dir) != dir_get_offset(last_collision))
(gdb)
590	    if (dir_probe(&key, vol, &dir, &last_collision)) {
(gdb)
591	      int ret = do_read_call(&key);
(gdb)
592	      if (ret == EVENT_RETURN)
(gdb)
594	      return EVENT_CONT;
(gdb) c
Continuing.

Breakpoint 2, CacheVC::openReadReadDone (this=0x7ffef001b360, event=3900, e=0x0) at CacheRead.cc:559
559	      if (doc->magic != DOC_MAGIC) {
(gdb) p *doc
$56 = {magic = 1595054867, len = 3888, total_len = 0, first_key = {b = {7936001979168651200, 1986278447822286449}, u64 = {7936001979168651200, 1986278447822286449}, u32 = {3643078592, 1847744448,
      2071342705, 462466489}, u8 = "\300\367$\331\300W\"nq.v{\271\255\220\033"}, key = {b = {18283213176734494251, 3565778776571952605}, u64 = {18283213176734494251, 3565778776571952605}, u32 = {11819,
      4256892292, 2905783773, 830222567}, u8 = "+.\000\000\204\005\273\375ݽ2\255\347\060|1"}, hlen = 3816, doc_type = 3, v_major = 24, v_minor = 0, unused = 0, sync_serial = 18929,
  write_serial = 815749, pinned = 0, checksum = 2695938256}
(gdb) n
573	      if (doc->key == key)
(gdb)
576	      else if (dir_ininterim(&dir)) {
(gdb)
583	    if (last_collision && dir_get_offset(&dir) != dir_get_offset(last_collision))
(gdb)
590	    if (dir_probe(&key, vol, &dir, &last_collision)) {
(gdb)
591	      int ret = do_read_call(&key);
(gdb)
592	      if (ret == EVENT_RETURN)
(gdb)
594	      return EVENT_CONT;
(gdb)
631	}
(gdb) c
Continuing.

Breakpoint 2, CacheVC::openReadReadDone (this=0x7ffef001b360, event=3900, e=0x0) at CacheRead.cc:559
559	      if (doc->magic != DOC_MAGIC) {
(gdb) p *doc
$57 = {magic = 1595054867, len = 3016, total_len = 4194304, first_key = {b = {7936001979168651200, 1986278447822286449}, u64 = {7936001979168651200, 1986278447822286449}, u32 = {3643078592, 1847744448,
      2071342705, 462466489}, u8 = "\300\367$\331\300W\"nq.v{\271\255\220\033"}, key = {b = {5241550030087605862, 11374957274309961022}, u64 = {5241550030087605862, 11374957274309961022}, u32 = {
      1070550630, 1220393467, 3061687614, 2648438623}, u8 = "fN\317?\373\271\275H>\245}\266_\367۝"}, hlen = 2944, doc_type = 3, v_major = 24, v_minor = 0, unused = 0, sync_serial = 34144,
  write_serial = 851504, pinned = 0, checksum = 2695938256}
(gdb) n
573	      if (doc->key == key)
(gdb)
576	      else if (dir_ininterim(&dir)) {
(gdb)
583	    if (last_collision && dir_get_offset(&dir) != dir_get_offset(last_collision))
(gdb)
590	    if (dir_probe(&key, vol, &dir, &last_collision)) {
(gdb)
595	    } else if (write_vc) {
(gdb)
619	  Warning("Document %s truncated", earliest_key.toHexStr(tmpstring));
(gdb)
620	  return calluser(VC_EVENT_ERROR);
(gdb)
631	}
(gdb)

**Comment by taoyunxing** [comment]
I try to explain the above debug procedure and why we get the 'Document truncated' log:
During getting the last fragment data, we encounter the voilent directory has collision(two ok, one fail):
First, we parse Doc, check doc->key != key, dir_probe, do_read_call, reentry CacheVC::openReadReadDone;
Second, we parse Doc, check doc->key != key, dir_probe, do_read_call, reentry CacheVC::openReadReadDone;   
Third,  we parse Doc, check doc->key != key, dir_probe, but this time we find no matched dir, finally, only just print 'Document truncated ……' log and go to error exception.

why we find no matched dir for the last fragment?

**Comment by bryancall** [comment]
@SolidWallOfCode can you comment on this?

**Comment by oknet** [comment]
In my opinion, we should revert the TS-3054 until the PR #2699 is reviewed or the TS-3054 is reimplemented. Because it makes the incomplete content is cached.

```
commit 05cfba6a77c48bf2f651e7d88202d5af3f676074
Author: Susan Hinrichs <shinrich@network-geographics.com>
Date:   Mon Sep 22 11:56:52 2014 -0500

    TS-3054: Pass truncated chunked data to client.
```

**Comment by taoyunxing** [comment]
I agree with @oknet

**Comment by scw00** [comment]
+1

**Comment by scw00** [comment]
this feature will be removed if #2865 accepted.

I'll open a new issue to discuss how to reimplement TS-3054.

According to TS-3054, what does ats do if content is not chunk ? 

It seems we also lost the initial body when server close immediately. 

```
package main

import (
	"net"
	"fmt"
)

const response =
	`HTTP/1.1 200 OK
Date: Sun, 31 Aug 2014 10:00:38 GMT
Server: Microsoft-IIS/6.0
X-Powered-By: ASP.NET
X-AspNet-Version: 4.0.30319
Content-Length: 1000000
Cache-Control: no-cache
Pragma: no-cache
Expires: -1
Content-Type: application/xml;; charset=utf-8

34
<recording><validtoken>true</validtoken></recording>`

func main()  {
	l, _ := net.Listen("tcp", ":12345")

	for true {
		c, _ := l.Accept()
		fmt.Println("revceived one connection")
		n, err := c.Write([]byte(response))
		fmt.Println(n, err, len(response))
		c.Close()
	}

}

```

@zwoop @shinrich @SolidWallOfCode

**Comment by zhxiaom5** [comment]
Maybe thats not just TS-3054 error，We use trafficserver 4.2.3，that version not apply TS-3054，but we also meet this error。
```
[Aug 14 07:06:06.128] Server {0x15225dfd3700} WARNING: Middle: Doc magic does not match for 00B59E24A25A0933711746E2DD2CF01C
[Aug 14 07:06:06.128] Server {0x15225dfd3700} WARNING: Document 69847794EB607DF271799448950874D6 truncated
[Aug 14 07:42:48.794] Server {0x15225d0c4700} WARNING: Head : Doc magic does not match for 7F4EA6372F3CFEF9C32301E20019337A
[Aug 14 08:18:46.911] Server {0x15225d9cd700} WARNING: Middle: Doc magic does not match for 00B59E24A25A0933711746E2DD2CF01C
[Aug 14 08:18:46.911] Server {0x15225d9cd700} WARNING: Document 69847794EB607DF271799448950874D6 truncated
[Aug 14 10:01:13.990] Server {0x15225dace700} WARNING: Head : Doc magic does not match for D5A882B58598E29B196A700F4D8426BC
[Aug 14 10:01:20.990] Server {0x15225d9cd700} WARNING: Middle: Doc magic does not match for 00B59E24A25A0933711746E2DD2CF01C
[Aug 14 10:01:20.990] Server {0x15225d9cd700} WARNING: Document 69847794EB607DF271799448950874D6 truncated
[Aug 14 10:24:53.279] Server {0x15225d7cb700} WARNING: Middle: Doc magic does not match for 00B59E24A25A0933711746E2DD2CF01C
[Aug 14 10:24:53.280] Server {0x15225d7cb700} WARNING: Document 69847794EB607DF271799448950874D6 truncated

< Age: 1548199
< Proxy-Connection: keep-alive
< 
{ [data not shown]
 18  131M   18 23.9M    0     0  23.7M      0  0:00:05  0:00:01  0:00:04 23.8M* transfer closed with 113170515 bytes remaining to read
 18  131M   18 23.9M    0     0  23.8M      0  0:00:05  0:00:01  0:00:04 23.8M* Closing connection #0

curl: (18) transfer closed with 113170515 bytes remaining to read
```

**Comment by bryancall** [comment]
If this is still an issue please reopen.
